Mathlib/Tactic/LieAlgebra/Basic.lean

```diff
@@ -0,0 +1,555 @@
+    | _ => els
+  | _ => els
+
+private theorem eq_aux {α} {a b c : α} (_ : (a : α) = c) (_ : b = c) : a = b := by subst_vars; rfl
+
+/-- Prove an equality in a `LieRing` by reducing two sides of the equation to Lyndon normal form. -/
+def proveEq (g : MVarId) : AtomM Unit := do
+  let some (α, e₁, e₂) := (← whnfR <|← instantiateMVars <|← g.getType).eq?
+    | throwError "lie_ring failed: not an equality"
+  let .sort u ← whnf (← inferType α) | unreachable!
+  let v ← (try u.dec catch _ => throwError "not a type{indentExpr α}")
+  have α : Q(Type v) := α
+  let sα ←
+    try Except.ok <$> synthInstanceQ q(LieRing $α)
+    catch e => pure (.error e)
+  have e₁ : Q($α) := e₁; have e₂ : Q($α) := e₂
+  let eq ← match sα with
+  | .ok sα => lieCore sα e₁ e₂
+  | .error e => throw e
+  g.assign eq
+where
+  /-- Reducing two side of equation to the normal form and determine whether they are equal. -/
+  lieCore {v : Level} {α : Q(Type v)} (sα : Q(LieRing $α))
+      (e₁ e₂ : Q($α)) : AtomM Q($e₁ = $e₂) := do
+    profileitM Exception "lie_ring" (← getOptions) do
+      let ⟨a, va, pa⟩ ← eval sα e₁
+      let ⟨_b, vb, pb⟩ ← eval sα e₂
+      unless va.eq vb do
+        throwError "tactic lie_ring failed, expressions are not equal, the left hand side is \
+        simplified to {a} but the right hand side is simplified to {_b}\n"
+      let pb : Q($e₂ = $a) := pb
+      return q(eq_aux $pa $pb)
+
+/--
+A tactic which evaluate an equality of two expressions in the `LieRing` to the Lyndon normal form,
+and check if they are equal.
+Notice that it only handle expressions consisting only of addition, subtraction, literal
+`ℤ` and `ℕ`-scalar multiplication, and Lie bracket.
+To prove an equality in an Lie algebra, please try `lie_algebra`.
+-/
+elab (name := lie_ring) "lie_ring" : tactic =>
+  withMainContext do
+    let s ← saveState
+    try
+      liftMetaMAtMain fun g ↦ do
+        AtomM.run .reducible (proveEq g)
+    catch e =>
+      restoreState s
+      throw e
+
+end execution
+
+section command
+
+/-- A Command which evaluates a `LieRing` expression to its Lyndon normal form. -/
+syntax (name := lie_reduce_cmd) "#lie_reduce" term : command
+
+open Command in
+@[command_elab lie_reduce_cmd] private def lieReduceCmdImpl :
+  Command.CommandElab :=
+  fun stx => withoutModifyingEnv <| runTermElabM fun _ => Term.withDeclName `_lie_reduce_cmd do
+    match stx with
+    | `(#lie_reduce $e) =>
+      try
+        let e ← Term.elabTerm e none
+        let α ← inferType e
+        Term.synthesizeSyntheticMVarsNoPostponing
+        let .sort u ← whnf (← inferType α) | unreachable!
+        let v ← try u.dec catch _ => throwError "not a type {indentExpr α}"
+        have α : Q(Type v) := α
+        let sα ← synthInstanceQ q(LieRing $α)
+        let ⟨a, _, _⟩ ← Mathlib.Tactic.AtomM.run .reducible (eval sα e)
+        -- TryThis.addTermSuggestion stx a
+        logInfo m!"the term is reduced to {a}"
+        return
```

Comment:
```suggestion
        let sα ← synthInstanceQ q(LieRing $α)
        let a ← Mathlib.Tactic.AtomM.run .reducible (eval sα e)
        -- TryThis.addTermSuggestion stx a
        logInfo m!"the term is reduced to {a.1}"
```
For some reason you get an error if you don't do this. Also please either use the comment or remove it.